msprime/simulations.py

```diff
@@ -1471,6 +1489,178 @@ def print_history(self, output=sys.stdout):
+        """
+        This function will calculate the mean coalescence rates between the
+        lineages of the sample configuration provided in `num_samples`, at
+        each generation time listed by `steps`, in this demographic model.
+        The coalescence rate at time t in the past is the average rate
+        of coalescence of as-yet-uncoalesed lineages, computed as follows:
+        let p(t) be the probability that the lineages of a randomly
+        chosen pair of samples has not yet coalesced by time t,
+        let p(z,t) be the probability that the lineages of a randomly
+        chosen pair of samples has not yet coalesced by time t *and* are
+        both in population z, and let N(z,t) be the diploid effective
+        population size of population z at time t. Then the mean coalescence
+        rate at time t is r(t) = (sum_z p(z,t) / (2 * N(z,t) ) / p(t).
+        The computation is done approximating population size trajetories
+        with piecewise constant trajectories, in steps of size dt = end / num_steps.
+        For this to be accurate, dt must be small enough so that (a) short epochs
+        (e.g., bottlenecks) are not missed, and (b) populations do not
+        change in size too much over that time, if they are growing or shrinking.
+        If in doubt, check that results are simliar after doubling `num_steps`.
+
+        :param list steps: The respective time steps (in generations) for which the
+            coalescence rate, as well as probablility that a randomly pair of lineages
+            has not yet coalesced, is computed and returned. The list given should be
+            sorted in ascending order.
+
+        :param list num_samples: is a list of the same length as the number
+            of populations, so that `num_samples[j]` is the number of sampled
+            chromosomes in subpopulation `j`.
+
+        :param int min_pop_size: is the smallest allowed population size during
+            computation of coalescent rates (i.e., coalescence rates are actually
+            1 / 2 * max(min_pop_size, N(z,t))). Spurious very small population sizes
+            can occurIn models where populations grow exponentially but are unused
+            before some time in the past, and lead to floating point error.
+            This should be set to a value smaller than the smallest
+            desired population size in the model.
+
+        :return: Returns a tuple where the first value is an array whose jth
+            element is the coalescence rate at the jth time point
+            (denoted r(t[j]) above), and the second is the probablility that a
+            randomly pair of lineages has not yet coalesced (denoted p(t[j]) above).
+
+        :rtype: class:'tuple'
+        """
+        if double_step_validatation:
+            offset = np.append(steps[1:], steps[-1]+1)
+            double_steps = [j for i in zip(steps, (steps+offset)/2) for j in i]
```

Comment:
This assumes that the last `+1` is smaller than the stepsize elsewhere. I'd just intercalculate one extra step in each spot, like so:
```
double_steps = sorted(np.concatenate([x[1:], x[1:] + np.diff(x)/2]))
```